src/centrality/truss.cpp

```diff
@@ -0,0 +1,271 @@
+ * then destroyed.
+ * \param truss This will indicate the highest k-truss each edge occurs in.
+ * \return Error code.
+ *
+ *
+ * Time complexity: It should be O(|E|^1.5). See Algorithm 2 in:
+ * Wang, Jia, and James Cheng. "Truss decomposition in massive networks."
+ * Proceedings of the VLDB Endowment 5.9 (2012): 812-823.
+ *
+ *
+ * \example test/unit/truss.cpp FIXME: can we make a C file instead?
+ */
+int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss){
+
+  igraph_integer_t fromVertex, toVertex, e1, e2;
+  igraph_vector_bool_t completed;
+  igraph_adjlist_t adjlist;
+  igraph_vector_int_t fromNeighbors, toNeighbors, q1, q2, commonNeighbors;
+  bool e1_complete, e2_complete;
+  int j, n, level, newLevel;
+  long int i, seed, ncommon;
+  long int nedges = igraph_vector_int_size(support);
+
+  // Get max possible value = max entry in support.
+  int max = igraph_vector_int_max(support);
+
+  // The vector of levels. Each level of the vector is a set of edges initially
+  // at that level of support, where support is # of triangles the edge is in.
+  vector< unordered_set<long int> > vec(max + 1);
+
+  // Initialize completed edges.
+  IGRAPH_VECTOR_BOOL_INIT_FINALLY(&completed, igraph_ecount(graph));
+
+  // Add each edge to its appropriate level of support.
+  for (i = 0; i < nedges; ++i) {
+    vec[VECTOR(*support)[i]].insert(i);  // insert edge i into its support level
+  }
+
+  // Record the trussness of edges at level 0. These edges are not part
+  // of any triangles, so there's not much to do and we "complete" them
+  unordered_set<long int>::iterator it;
+  for (it = vec[0].begin(); it != vec[0].end(); ++it){
+    VECTOR(*truss)[*it] = 0;
+    VECTOR(completed)[*it] = 1;
+  }
+
+  // Initialize variables needed below.
+  /* FIXME: I think multiedges but not loops should stay here? */
+  IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_ALL, IGRAPH_NO_LOOPS,
+                                   IGRAPH_MULTIPLE));
+  IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+  IGRAPH_VECTOR_INT_INIT_FINALLY(&commonNeighbors, 0);
+
+  /* Sort each vector of neighbors, so it's done once for all */
+  igraph_adjlist_sort(&adjlist);
+
+  // Move through the levels, one level at a time, starting at first level.
+  for (level = 1; level <= max; ++level){
+
+    /* Track down edges one at a time */
+    while (!vec[level].empty()){
+      seed = *vec[level].begin();  // pull out the first edge
+      vec[level].erase(seed);  // remove the first element
+
+      /* Find the vertices of this edge */
+      igraph_edge(graph, seed, &fromVertex, &toVertex);
+
+      /* Find neighbors of both vertices. If they run into each other,
+       * there is a triangle. Because we sorted the adjacency list already,
+       * we don't need to sort it for every edge here */
+      fromNeighbors = adjlist.adjs[fromVertex];
+      toNeighbors = adjlist.adjs[toVertex];
```

Comment:
Let's stick to the documented API and use `igraph_adjlist_get` here instead of directly accessing the innards of the the adjlist.